### PR TITLE
Implement NetworkAddressChanged event on OSX

### PR DESCRIPTION
**Pull request by mellinoe**
This is the last work item I have for NetworkInformation. It addresses Issue #4058 .

I'm using the OSX SystemConfiguration Framework to listen for notifications about network address changes. The implementation here works, but I'd like some feedback on the threading approach and the general style I've used for cleaning up the temporary OSX string objects that need to be created for the native calls. It's fairly messy right now.

@sokket I followed your general example from FileSystemWatcher and adapted it a bit for this.

Please take a look, @sokket, @stephentoub , @CIPop

### CODE REVIEW

File: src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs [hunk 1]
```diff
@@ -1,22 +1,183 @@
...
+
+        // Listener thread that adds the RunLoopSource to its RunLoop.
+        private static Thread s_runLoopThread;
+
+        // The listener thread's CFRunLoop.
+        private static CFRunLoopRef s_runLoop;
+
+        // Use an event to try to prevent StartRaisingEvents from returning before the
+        // RunLoop actually begins. This will mitigate a race condition where the watcher
+        // thread hasn't completed initialization and stop is called before the RunLoop even starts.
```
Comment (stephentoub) [inline src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs]: Nit: even => event

**Comment by sokket** [comment]
LGTM, nicely done!

File: src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs (hunk 1 above)
Comment (mellinoe) [inline src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs]: I think it's correct, if you're referring to the second-last word. Perhaps I should reword it to "before the run loop has even started". Or just remove "even" :smile:

File: src/Common/src/Interop/OSX/Interop.SystemConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            IntPtr info);
+
+        /// <summary>
+        /// Creates a new session used to interact with the dynamic store maintained by the System Configuration server.
+        /// Follows the "Create Rule" where if you create it, you delete it.
+        /// </summary>
+        /// <param name="allocator">Should be IntPtr.Zero</param>
+        /// <param name="name">The name of the calling process or plug-in of the caller.</param>
+        /// <param name="callout">The function to be called when a watched value in the dynamic store is changed.
+        /// Pass NULL if no callouts are desired.</param>
```
Comment (mellinoe) [inline src/Common/src/Interop/OSX/Interop.SystemConfiguration.cs]: Will change "NULL" to "IntPtr.Zero"

File: src/Common/src/Interop/OSX/Interop.SystemConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            SCDynamicStoreCallBack callout,
+            SCDynamicStoreContext* context);
+
+        /// <summary>
+        /// Creates a new session used to interact with the dynamic store maintained by the System Configuration server.
+        /// Follows the "Create Rule" where if you create it, you delete it.
+        /// </summary>
+        /// <param name="name">The name of the calling process or plug-in of the caller.</param>
+        /// <param name="callout">The function to be called when a watched value in the dynamic store is changed.
+        /// Pass NULL if no callouts are desired.</param>
```
Comment (mellinoe) [inline src/Common/src/Interop/OSX/Interop.SystemConfiguration.cs]: NULL -> IntPtr.Zero

File: src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs [hunk 4]
```diff
@@ -1,22 +1,183 @@
...
+        private static unsafe void CreateAndStartRunLoop()
+        {
+            Debug.Assert(s_dynamicStoreRef == null);
+
+            var storeContext = new Interop.SystemConfiguration.SCDynamicStoreContext();
+            using (SafeCreateHandle storeName = Interop.CoreFoundation.CFStringCreateWithCString("NetworkAddressChange.OSX"))
+            {
+                s_dynamicStoreRef = Interop.SystemConfiguration.SCDynamicStoreCreate(
+                    storeName.DangerousGetHandle(),
+                    OnAddressChanged,
```
Comment (stephentoub) [inline src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs]: This doesn't look safe.  The C# compiler is going to allocate a delegate, but that'll be collected as soon as this call to create returns, at which point the native could try to call back to a collected delegate.  This delegate should be stored into a static, and then that static passed in here.

File: src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs]: Ah, sorry, you're right, I misread it.

File: src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs (hunk 4 above)
Comment (mellinoe) [inline src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs]: Interesting, I must have never triggered a GC when testing this I suppose.

Comment (mellinoe) [inline src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.OSX.cs]: Should be fixed now, using a static field to store the delegate.

**Comment by mellinoe** [comment]
Any other comments, @stephentoub ? Does the thread usage seem reasonable other than the delegate issue?

**Comment by stephentoub** [comment]
LGTM
